### ISSUE TITLE
We need batch processing

### ISSUE THREAD

**Issue by e1732a364fed** [ml-explore/mlx-lm#178]
See

https://github.com/ml-explore/mlx-examples/issues/1268

https://github.com/ml-explore/mlx-examples/pull/948

**Comment by otarkhan** [comment]
@awni

**Comment by jgburford** [comment]
I read through prior discussions and saw no mention of https://github.com/nath1295/MLX-Textgen

Seems to already have batch processing implemented using mlx, so perhaps it would be feasible to adapt it to mlx-lm

**Comment by ghost** [comment]
> I read through prior discussions and saw no mention of https://github.com/nath1295/MLX-Textgen
> 
> Seems to already have batch processing implemented using mlx, so perhaps it would be feasible to adapt it to mlx-lm

Wow Thanks! I tried it and it worked!

with a little effort: edit `mlx_textgen/model_utils.py` and change the "convert" import to `from mlx_lm import convert` 

and run with ` --model-path` instead of `--model`.

It's still slower than vllm ( way slower )... but worth a try. 

But during normal tasks,  it seems that (MLX-Textgen and mlx_parallel) didn't speed things up.

**Comment by nath1295** [comment]
> I read through prior discussions and saw no mention of https://github.com/nath1295/MLX-Textgen
> 
> Seems to already have batch processing implemented using mlx, so perhaps it would be feasible to adapt it to mlx-lm



> > I read through prior discussions and saw no mention of https://github.com/nath1295/MLX-Textgen
> > Seems to already have batch processing implemented using mlx, so perhaps it would be feasible to adapt it to mlx-lm
> 
> Wow Thanks! I tried it and it worked!
> 
> with a little effort: edit `mlx_textgen/model_utils.py` and change the "convert" import to `from mlx_lm import convert`
> 
> and run with ` --model-path` instead of `--model`.
> 
> It's still slower than vllm ( way slower )... but worth a try.

Hi there, Im the guy who made mlx-textgen. Although batch processing is kind of working with it, I did not take care of the attention mask with paddings properly, which might lead to inconsistent results during batch processing even with the same prompt. 

With the latest updates of mlx-lm, I am trying to implement the mask properly, but it seems that there is something wrong with `mlx.fast.scaled_dot_product_attention`. Even with the exact same prompt in the same batch, the model gives me two completely different sets of logits for the next token per prompt. That's even before applying mask explicitly (Same prompt in the batch, means no padding needed, so I simply pass mask=None). @awni might have a better explanation to this?

**Comment by jgburford** [comment]
Here's another mlx based inference implementation I found no mention of in discussions/issues in the mlx-lm github https://github.com/GreenBitAI/gbx-lm 

It says it can handle multiple requests concurrently, but no mention of the actual terms batching nor continuous batching.

It says it has optional compatibility with existing mlx-community / mlx-lm models but mainly targets a different quant format developed internally by Greenbit

**Comment by awni** [comment]
@nath1295 regarding:

> With the latest updates of mlx-lm, I am trying to implement the mask properly, but it seems that there is something wrong with mlx.fast.scaled_dot_product_attention. Even with the exact same prompt in the same batch, the model gives me two completely different sets of logits for the next token per prompt. 

If you can provide a test case to reproduce the issue we can take a look and make sure it's not a bug in SDPA or mlx-lm.

**Comment by nath1295** [comment]
@awni Here is a simple test script I have with qwen2.5 coder 3B, you just need to change the `MODEL_DIR` to load the model.

```python
from typing import List, Dict, Union, TYPE_CHECKING
if TYPE_CHECKING:
    from mlx.core import array
    from mlx.nn import Module
    from mlx_lm.models.cache import KVCache, RotatingKVCache
    from transformers.tokenization_utils import PreTrainedTokenizer


class ModelTest:

    def __init__(self, model_dir: str):
        from mlx_lm.utils import load_model
        from transformers import AutoTokenizer
        from pathlib import Path

        model, config = load_model(Path(model_dir))
        self.model: "Module" = model
        self.tokenizer: "PreTrainedTokenizer" = AutoTokenizer.from_pretrained(model_dir)

    def get_empty_cache(self) -> List[Union["KVCache", "RotatingKVCache"]]:
        from mlx_lm.models.cache import make_prompt_cache
        return make_prompt_cache(self.model)

    def tokenize(self, prompts: List[str]) -> Dict[str, "array"]:
        return self.tokenizer(prompts, return_tensors='mlx', padding=True, padding_side='left')
    
    def get_next_token_logits(self, prompts: List[str]) -> "array":
        inputs = self.tokenize(prompts)
        return self.model(inputs['input_ids'])
    
    def get_embeddings(self, prompts: List[str]) -> "array":
        inputs = self.tokenize(prompts)
        embed = self.model.model.embed_tokens
        return embed(inputs['input_ids'])
    
    def get_mask(self, h: "array") -> "array":
        from mlx_lm.models.base import create_attention_mask
        return create_attention_mask(h)
    
    def get_embeddings_first_layer(self, prompts: List[str]) -> "array":
        embeddings = self.get_embeddings(prompts)
        mask = self.get_mask(embeddings)
        cache = self.get_empty_cache()
        c = cache[0]
        layer = self.model.model.layers[0]
        return layer(embeddings, mask, c)


if __name__ == '__main__':

    # Using "mlx-community/Qwen2.5-Coder-3B-Instruct-8bit" for this test
    MODEL_DIR = 'My/Path/To/Qwen2.5CoderModel'

    test_model = ModelTest(MODEL_DIR)

    prompts = [
        'Hi there',
        'Hi there'
    ]

    # Get the next token logits for each prompt first
    print('Model last token logits for the two prompts:')
    print('-' * 100)
    next_token_logits = test_model.get_next_token_logits(prompts)[... , -1, :]
    print(next_token_logits) # Different next token logits
    print('-' * 100)

    # Now look at the embeddings before entering the transformer blocks to check if everything is okay
    print('\nEmbeddings before transformer blocks:')
    print('-' * 100)
    print(test_model.get_embeddings(prompts)) # Everything looks good here
    print('-' * 100)

    # Now look at the embeddings after the first transformer block
    print('\nEmbeddings after first transformer block:')
    print('-' * 100)
    print(test_model.get_embeddings_first_layer(prompts)) # Different embeddings for the two prompts, so weird stuff happening here
    print('-' * 100)


    print('Done.')
```

In my environment, I have `mlx-lm==0.24.0` and `mlx==0.25.1` installed. Let me know if I did something wrong.

**Comment by ghost** [comment]
I take back my "positive review" for MLX-Textgen and mlx_parallm.

I tested them with professional benchmark tool, and realized that they did not speed up at all.

I used this tool:
 https://github.com/huggingface/inference-benchmarker

To test mlx_parallm I used this fork: https://github.com/misanthropic-ai/mlx_parallm

I just leave the link here without releasing my benchmark result, because maybe I was wrong somewhere.

My result is that MLX-Textgen or mlx_parallm is just slower than using mlx-lm directly, 

The speed (total throughput) of mlx-lm is only about 1/50 of vllm( running in linux with a nvidia gpu, as vllm doesn't support metal yet.), approximately, as mac and nvidia gpus are so different. And this throughput is just slower than speed of a single request, so no parallel at all apparently.

(and MLX-Textgen just error out when running the benchmark where vllm, mlx_parallm fork and lm-studio(which uses mlx-lm) are all ok )



 People can test them to verify.

**Comment by otarkhan** [comment]
@e1732a364fed I'm currently testing with https://github.com/EricLBuehler/mistral.rs, give it a shot

**Comment by ghost** [comment]
> [@e1732a364fed](https://github.com/e1732a364fed) I'm currently testing with https://github.com/EricLBuehler/mistral.rs, give it a shot

I tried it. It does support mlx, but it still doesn't look like with any parallel processing (on mac).

**Comment by otarkhan** [comment]
Actually it kinda does. With server mode and `--max-seqs` > 1 it'll be processing multiple requests in parallel

**Comment by ghost** [comment]
> Actually it kinda does. With server mode and `--max-seqs` > 1 it'll be processing multiple requests in parallel

I tested with server mode and default --max-seqs (16)

What's your inference-benchmarker result throughput then? I tested it and it doesn't look good.

It does show some "Prefix cache hitrate 97.50%," logs in mistral.rs, but in benchmark, it's still perform slower (on average) than a single request

**Comment by ghost** [comment]
> Actually it kinda does. With server mode and `--max-seqs` > 1 it'll be processing multiple requests in parallel

Ok，it looks like that if I enable "paged-attention", it might have got some speed up. (But did not, in fact.)

However, I just met you in this issue
https://github.com/EricLBuehler/mistral.rs/issues/1348#issuecomment-2935521954

so... We'd better talk about mistral.rs there...

**Comment by ghost** [comment]
I tested MLX-Textgen / mlx_parallm / vllm again with llmperf

https://github.com/ray-project/llmperf


MLX-Textgen  and mlx_parallm still have no speed up at all for concurrent-requests. And vllm is still about 10x faster with concurrent-requests  than single request.

People can test them to verify.

vllm is still the best option, it's a loss that it does not support metal. That's why we struggling here.

**Comment by MerrillLi** [comment]
> Here's another mlx based inference implementation I found no mention of in discussions/issues in the mlx-lm github https://github.com/GreenBitAI/gbx-lm
> 
> It says it can handle multiple requests concurrently, but no mention of the actual terms batching nor continuous batching.
> 
> It says it has optional compatibility with existing mlx-community / mlx-lm models but mainly targets a different quant format developed internally by Greenbit

I looked into their code... actually, it is more like the concurrency when serving web requests --- they usually start up a thread for each incoming request.

**Comment by ivanfioravanti** [comment]
This is extremely interesting! I'm playin with it to speed up evals and it seems working! Any news on this @awni ?

**Comment by awni** [comment]
Yes, we definitely want to have better support for batching. It will take some time to get there, but it's on the roadmap.

**Comment by otarkhan** [comment]
@awni if that helps, someone already did some effort there but the repo hasn't been maintained for a while now.
https://github.com/willccbb/mlx_parallm/tree/main

**Comment by ivanfioravanti** [comment]
I'm working/testing mlx_parallm, but there are many missing part there and KL Divergence is quite big. But it's a great starting point.

**Comment by do-me** [comment]
Thought I'd just chime in with some plots and benchmarks I ran a while ago: https://github.com/willccbb/mlx_parallm/pull/5
I got some very impressive performance back then, around 700 tokens/s with mlx-community/Meta-Llama-3-8B-Instruct-4bit on an M3 Max.

**Comment by awni** [comment]
We can close this as it's supported in the latest mlx-lm.

The `batch_generate` function is the place to start. See the [example](https://github.com/ml-explore/mlx-lm/blob/main/mlx_lm/examples/batch_generate_response.py)
